sympy/algebras/dual.py

```diff
@@ -0,0 +1,338 @@
+# References :
+# https://en.wikipedia.org/wiki/Dual_number
+from sympy import conjugate, exp, ln, sympify
+from sympy.core.expr import Expr
+from sympy.algebras.quaternion import Quaternion
+
+
+class Dual(Expr):
+    """Provides basic dual number operations. Dual objects can be instantiated
+    as Dual(a, b), as in (a + b*d), where d^2 = 0.
+
+    Dual numbers can be initiated with Quaternions to create Dual Quaternions,
+    the algebra of all rigid body kinematics in 3D. In other words, translations
+    and rotations. Another important use of dual numbers is automatic
+    differentiation.
+
+    Examples
+    ========
+
+    Dual Quaternions are created by initiating a dual number from quaternions:
+
+    >>> a, b, c, d, e, f, g, h = symbols('a:h')
+    >>> p = Quaternion(a, b, c, d)
+    >>> q = Quaternion(e, f, g, h)
+    >>> x = Dual(p, q)
+    >>> x
+    a + b*i + c*j + d*k + (e + f*i + g*j + h*k)*ð
+    >>> x*x.inverse()
+    1 + 0*i + 0*j + 0*k + (0 + 0*i + 0*j + 0*k)*ð
+
+    Take note that operations like conjugate only perform the dual conjugate,
+    not the quaternionic conjugate.
+
+    >>> conjugate(x)
+    a + b*i + c*j + d*k + ((-e) + (-f)*i + (-g)*j + (-h)*k)*ð
+
+    As an example of automatic differentiation consider the simple function x^2.
+    Evaluating the function with a dual number automatically performs the chain
+    rulewhile the function itself is being evaluated.
+
+    >>> from sympy import symbols
+    >>> from sympy.algebras.dual import Dual
+    >>> a = symbols('a')
+    >>> d = Dual(a, 1)
+    >>> d
+    a + 1*d
+    >>> f = lambda x: x**2
+    >>> f(d)
+    a**2 + 2*a*d
+
+    We recognize the term proportional to d as the derivative of the function.
+    To obtain only the derivative we select only the dual component:
+
+    >>> f(d).b
+    2*a
+    """
+    _op_priority = 11.0
+
+    is_commutative = True
+
+    def __new__(cls, a=0, b=0):
+        a = sympify(a)
+        b = sympify(b)
+
+        obj = Expr.__new__(cls, a, b)
+        obj._a = a
+        obj._b = b
+        if isinstance(a, Quaternion) and isinstance(b, Quaternion):
+            # Dual quaterions are not a commutative algebra
+            obj.is_commutative = False
+        return obj
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
```

Comment:
Idiomatically it seems that
```suggestion
    @property
    def a(self):
        return self.args[0]

    @property
    def b(self):
        return self.args[1]
```
is how sympy handles this